mysql/infoschema.go

```diff
@@ -304,6 +310,78 @@ func getConstraints(conv *internal.Conv, db *sql.DB, table schemaAndName) ([]str
 	return primaryKeys, m, nil
 }
 
+// getForeignKeys return list all the foreign keys constraints.
+func getForeignKeys(conv *internal.Conv, db *sql.DB, table schemaAndName) (foreignKeys []schema.ForeignKey, err error) {
+	refTables, err := getRefTables(db, table)
```

Comment:
Why do we need to getRefTables?
Couldn't we get this from INFORMATION_SCHEMA.KEY_COLUMN_USAGE?

In other words, couldn't we call getForeignKey with just db and table, and when we join the INFORMATION_SCHEMA.TABLE_CONSTRAINTS and INFORMATION_SCHEMA.KEY_COLUMN_USAGE tables, we could add  k.REFERENCED_TABLE_NAME to the select?

Reply:
Yes Nevin, you pointed it out correctly. We need to collect cols & referCols list by Table/RefTable Pair. That's why I came up with this approach. 

The problem with the query would be that it will return redundant rows eg:
```
+-----------------------+----------------------+------------------------+
| REFERENCED_TABLE_NAME | COLUMN_NAME          | REFERENCED_COLUMN_NAME |
+-----------------------+----------------------+------------------------+
| actor                 | actor_id             | actor_id               |
| address               | address_id           | address_id             |
| address               | address_id           | address_id             |
| address               | address_id           | address_id             |
| category              | category_id          | category_id            |
| city                  | city_id              | city_id                |
| country               | country_id           | country_id             |
| customer              | customer_id          | customer_id            |
| customer              | customer_id          | customer_id            |
| film                  | film_id              | film_id                |
| film                  | film_id              | film_id                |
| film                  | film_id              | film_id                |
+-----------------------+----------------------+------------------------+
```

Now we can argue that we can use `distinct`, but it will throw an error like this:

```
ERROR 3065 (HY000): Expression #2 of ORDER BY clause is not in SELECT list, references column 'k.ORDINAL_POSITION' which is not in SELECT list; this is incompatible with DISTINCT
```

Because of the above error, we also need to add k.ORDINAL_POSITION to the select query (But it will not have any significance). 

So, the final query would look like this:

```
SELECT distinct k.REFERENCED_TABLE_NAME,k.COLUMN_NAME,k.REFERENCED_COLUMN_NAME, k.ordinal_position 
		FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS t 
		INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS k 
			ON t.CONSTRAINT_NAME = k.CONSTRAINT_NAME 
			AND t.CONSTRAINT_SCHEMA = k.CONSTRAINT_SCHEMA 
			AND t.TABLE_NAME = k.TABLE_NAME 
		WHERE k.TABLE_SCHEMA = ? 
			AND k.TABLE_NAME = ? 
			AND t.CONSTRAINT_TYPE = "FOREIGN KEY" 
		ORDER BY
			k.REFERENCED_TABLE_NAME,
			k.ordinal_position;
```

and then we will use map of referTable & Cols & referCols.

If you think this is fine, then please let us know. So we will update our approach.

Reply:
I went ahead and implemented this approach for postgres and mysql both. But I have commented the previous approach (so it would be easier to roll back).  Please take a look and before merging this PR, we can remove previous approach or this approach.

Reply:
Many thanks!

I'm still confused on one point. You mentioned the issue of redundant rows and gave the example from sakila. But isn't that because you omitted TABLE_NAME from the query?

Here's my reconstruction of the table you gave with the redundant rows.

SELECT k.REFERENCED_TABLE_NAME,k.COLUMN_NAME,k.REFERENCED_COLUMN_NAME
		FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS t 
		INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS k 
			ON t.CONSTRAINT_NAME = k.CONSTRAINT_NAME 
			AND t.CONSTRAINT_SCHEMA = k.CONSTRAINT_SCHEMA 
			AND t.TABLE_NAME = k.TABLE_NAME 
			AND k.REFERENCED_TABLE_SCHEMA = k.TABLE_SCHEMA
		WHERE k.TABLE_SCHEMA = 'sakila' 
			AND t.CONSTRAINT_TYPE = "FOREIGN KEY" 
		ORDER BY
			k.REFERENCED_TABLE_NAME,
			k.ordinal_position;

which gives the following results (matching the table you gave):

```
+-----------------------+----------------------+------------------------+
| REFERENCED_TABLE_NAME | COLUMN_NAME          | REFERENCED_COLUMN_NAME |
+-----------------------+----------------------+------------------------+
| actor                 | actor_id             | actor_id               |
| address               | address_id           | address_id             |
| address               | address_id           | address_id             |
| address               | address_id           | address_id             |
| category              | category_id          | category_id            |
| city                  | city_id              | city_id                |
| country               | country_id           | country_id             |
| customer              | customer_id          | customer_id            |
| customer              | customer_id          | customer_id            |
| film                  | film_id              | film_id                |
| film                  | film_id              | film_id                |
| film                  | film_id              | film_id                |
```

But if you add in TABLE_NAME:

SELECT k.TABLE_NAME,k.REFERENCED_TABLE_NAME,k.COLUMN_NAME,k.REFERENCED_COLUMN_NAME
		FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS t 
		INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS k 
			ON t.CONSTRAINT_NAME = k.CONSTRAINT_NAME 
			AND t.CONSTRAINT_SCHEMA = k.CONSTRAINT_SCHEMA 
			AND t.TABLE_NAME = k.TABLE_NAME 
			AND k.REFERENCED_TABLE_SCHEMA = k.TABLE_SCHEMA
		WHERE k.TABLE_SCHEMA = 'sakila' 
			AND t.CONSTRAINT_TYPE = "FOREIGN KEY" 
		ORDER BY
			k.REFERENCED_TABLE_NAME,
			k.ordinal_position;

```
 +---------------+-----------------------+----------------------+------------------------+
| TABLE_NAME    | REFERENCED_TABLE_NAME | COLUMN_NAME          | REFERENCED_COLUMN_NAME |
+---------------+-----------------------+----------------------+------------------------+
| film_actor    | actor                 | actor_id             | actor_id               |
| staff         | address               | address_id           | address_id             |
| customer      | address               | address_id           | address_id             |
| store         | address               | address_id           | address_id             |
| film_category | category              | category_id          | category_id            |
| address       | city                  | city_id              | city_id                |
| city          | country               | country_id           | country_id             |
| payment       | customer              | customer_id          | customer_id            |
| rental        | customer              | customer_id          | customer_id            |
| film_actor    | film                  | film_id              | film_id                |
| film_category | film                  | film_id              | film_id                |
| inventory     | film                  | film_id              | film_id                |
```

and shows the source of the redundancy: it is foreign keys for different tables.

Or put another way, once you specify TABLE_NAME and TABLE_SCHEMA, the redundancy goes away. Hence we can get rid of the use of distinct and adding k.ordinal_position to the selected columns, and just use the query:

SELECT k.REFERENCED_TABLE_NAME,k.COLUMN_NAME,k.REFERENCED_COLUMN_NAME 
		FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS AS t 
		INNER JOIN INFORMATION_SCHEMA.KEY_COLUMN_USAGE AS k 
			ON t.CONSTRAINT_NAME = k.CONSTRAINT_NAME 
			AND t.CONSTRAINT_SCHEMA = k.CONSTRAINT_SCHEMA 
			AND t.TABLE_NAME = k.TABLE_NAME 
			AND k.REFERENCED_TABLE_SCHEMA = k.TABLE_SCHEMA
		WHERE k.TABLE_SCHEMA = ? 
			AND k.TABLE_NAME = ? 
			AND t.CONSTRAINT_TYPE = "FOREIGN KEY" 
		ORDER BY
			k.REFERENCED_TABLE_NAME,
			k.ordinal_position;

Reply:
This makes sense, i have changed the query at both the places in postgres and mysql.